optuna/storages/cached_storage.py

```diff
@@ -0,0 +1,291 @@
+from collections import defaultdict
+import copy
+import threading
+from typing import Any
+from typing import DefaultDict
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Set
+from typing import Tuple
+
+from optuna import distributions
+from optuna.storages import base
+from optuna.storages.rdb.storage import RDBStorage
+from optuna.study import StudyDirection
+from optuna.study import StudySummary
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+class _CachedStorage(base.BaseStorage):
```

Comment:
[Opinion Wanted] How about making this class public?
Currently, the `RDBStorage` instance created via the `get_storage` function will be wrapped by `_CachedStorage`, but users can create `RDBStorage` instance directly. It sometimes required to pads DB options such as `pool_pre_ping` and `wait_timeout` for MySQL (c.f., #622) . In such cases, users may want to apply cache to the `RDBStorage` instance by themselves.

Reply:
Ah, `get_storage` wraps `RDBStorage` and `get_storage` is called by `create_study`. So, `RDBStorage` will be wrapped by `_CachedStorage` and we can keep it private.